internal/lang/evalchecks/eval_for_each.go

```diff
@@ -225,6 +225,7 @@ func performSetValueChecks(expr hcl.Expression, hclCtx *hcl.EvalContext, forEach
 
 	// A set of strings may contain null, which makes it impossible to
 	// convert to a map, so we must return an error
+	forEachVal, _ = forEachVal.UnmarkDeep()
```

Comment:
I don't think it's correct to return the unmarked version of this value, because the caller of this function then passes the returned value on to another function:

https://github.com/opentofu/opentofu/blob/131b1456cae9c3359148c6b93b909b880e1bd739/internal/lang/evalchecks/eval_for_each.go#L105-L109

...and then that function wants to reject attempts to use a sensitive value:

https://github.com/opentofu/opentofu/blob/131b1456cae9c3359148c6b93b909b880e1bd739/internal/lang/evalchecks/eval_for_each.go#L172-L185

To make sure this change can't affect anything beyond fixing the bug as reported, I'd suggest that we instead change the following line to `it := forEachVal.UnmarkDeep().ElementIterator()`, so that the mark will still be present in all of the other places it previously was.

Reply:
Fixed in [634ca05](https://github.com/opentofu/opentofu/pull/3070/commits/634ca0570a2e6cc23435376f0d3b803a322e8ec9)